homework2/task01/project1/task1.py

```diff
@@ -43,26 +39,27 @@ def get_rarest_char(file_path: str) -> str:
     """Return one single rarest char."""
     with open(file_path, "r", encoding="unicode_escape") as f:
         text = f.read()
-        di = {}
+        all_chars = {}
         for i in text:
-            if i not in di:
-                di[i] = 1
+            if i not in all_chars:
+                all_chars[i] = 1
             else:
-                di[i] += 1
-        for key, value in di.items():
-            if value == min(di.values()):
+                all_chars[i] += 1
+        min_value = min(all_chars.values())
+        for key, value in all_chars.items():
+            if value == min_value:
                 return key
 
 
 def count_punctuation_chars(file_path: str) -> int:
     """Return the amount of punctuation chars."""
     with open(file_path, "r", encoding="unicode_escape") as f:
         text = f.read()
-        counter = 0
+        all_punctuation = dict.fromkeys(set(string.punctuation), 0)
```

Comment:
Does Python have another collection that allows checking the presence of the element in O(1)?

Reply:
> because elements in set are not ordered and there is no need to run through all n elements like in lists or tuples.

OrderedDict has order and O(1) search.  These conditions are not related. 

I'd say `set` and `int` (single counter) will do a better job here when dict.